File: mobile/lib/tags.dart
const maxCardTags = 10;
const maxTagLength = 30;

String normalizeTag(String value) =>
    value.trim().replaceAll(RegExp(r'\s+'), ' ');

({List<String> tags, String? error}) addCardTag(
  List<String> tags,
  String value,
) {
  final tag = normalizeTag(value);
  if (tag.isEmpty) return (tags: tags, error: 'Etiket boş olamaz.');
  if (tag.length > maxTagLength) {
    return (tags: tags, error: 'Etiket en fazla 30 karakter olabilir.');
  }
  if (tags.length >= maxCardTags) {
    return (tags: tags, error: 'Bir karta en fazla 10 etiket eklenebilir.');
  }
  if (tags.any((current) => current.toLowerCase() == tag.toLowerCase())) {
    return (tags: tags, error: 'Bu etiket zaten ekli.');
  }
  return (tags: [...tags, tag], error: null);
}

int tagColorIndex(String tag) {
  var hash = 0;
  for (final rune in tag.runes) {
    hash = (hash * 31 + rune) & 0x7fffffff;
  }
  return hash % 8;
}

File: web/src/lib/tags.ts
export const MAX_CARD_TAGS = 10;
export const MAX_TAG_LENGTH = 30;

export const normalizeTag = (value: string): string =>
  value.normalize('NFKC').trim().replace(/\s+/g, ' ');

export const tagKey = (value: string): string =>
  normalizeTag(value).replace(/[Iİ]/g, 'i').toLowerCase();

export function addTag(tags: string[], value: string): { tags: string[]; error?: string } {
  const tag = normalizeTag(value);
  if (tag.length === 0) return { tags, error: 'Etiket boş olamaz.' };
  if (tag.length > MAX_TAG_LENGTH) return { tags, error: 'Etiket en fazla 30 karakter olabilir.' };
  if (tags.length >= MAX_CARD_TAGS) return { tags, error: 'Bir karta en fazla 10 etiket eklenebilir.' };
  if (tags.some((current) => tagKey(current) === tagKey(tag))) {
    return { tags, error: 'Bu etiket zaten ekli.' };
  }
  return { tags: [...tags, tag] };
}

export function tagColorIndex(tag: string): number {
  let hash = 0;
  for (const character of tag) hash = (hash * 31 + character.codePointAt(0)!) >>> 0;
  return hash % 8;
}
